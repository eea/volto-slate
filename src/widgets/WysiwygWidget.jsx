/**
 * WysiwygWidget container.
 * @module volto-slate/widgets/WysiwygWidget
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';

import { FormFieldWrapper } from '@plone/volto/components';

import SlateEditor from '../editor/SlateEditor';
import { serializeNodesToHtml } from '../editor/render';
import { normalizeBlockNodes } from 'volto-slate/utils';
import { deserialize } from 'volto-slate/editor/deserialize';
import { Editor } from 'slate';

import './style.css';

/**
 * WysiwygWidget container class.
 * @class WysiwygWidget
 * @extends Component
 */
class WysiwygWidget extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    /**
     * Id of the field
     */
    id: PropTypes.string.isRequired,
    /**
     * Value of the field
     */
    value: PropTypes.shape({
      /**
       * Content type of the value
       */
      'content-type': PropTypes.string,
      /**
       * Data of the value
       */
      data: PropTypes.string,
      /**
       * Encoding of the value
       */
      encoding: PropTypes.string,
    }),
    /**
     * On change handler
     */
    onChange: PropTypes.func,
  };

  /**
   * Default properties
   * @property {Object} defaultProps Default properties.
   * @static
   */
  static defaultProps = {
    value: {
      'content-type': 'text/html',
      data: '',
      encoding: 'utf8',
    },
    onChange: null,
  };

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Component properties
   * @constructs WysiwygWidget
   */
  constructor(props) {
    super(props);

    this.state = {
      selected: this.props.focus,
      value: null,
    };

    this.editorRef = React.createRef(null);
  }

  /**
   * Change handler
   * @method onChange
   * @returns {undefined}
   */
  onChange = (data) => {
    this.setState(
      (state, props) => {
        return { ...state, value: data };
      },
      () => {
        if (this.unmounting) return;

        this.props.onChange(this.props.id, {
          data: serializeNodesToHtml(data),
        });
      },
    );
  };

  convertHTMLToNodes = (editor, html) => {
    return new Promise((resolve, reject) => {
      let rr = null;
      let nodes = [];
      this.setState(
        (state, props) => {
          if (rr !== null) {
            resolve([]);
            return state;
          }

          rr = Editor.rangeRef(editor, editor.selection, {
            affinity: 'forward',
          });

          const parsed = new DOMParser().parseFromString(html, 'text/html');

          let fragment = deserialize(editor, parsed.body);
          fragment = Array.isArray(fragment) ? fragment : [fragment];
          nodes = normalizeBlockNodes(editor, fragment);

          return { ...state, value: nodes };
        },
        () => {
          if (this.unmounting) {
            reject();
            return;
          }
          if (rr === null) {
            resolve([]);
            return;
          }
          // TODO: pt should be obtained from the rr above and passed to Transforms.select
          // call commented below.
          const pt = /* rr?.current || Editor.start(editor) ||  */ {};
          pt.anchor = { path: [0, 0], offset: 0 };
          pt.focus = { path: [0, 0], offset: 0 };
          // Transforms.select(editor, /* html.length === 0 ?  */pt/*  : rr.current */);
          rr.unref();
          rr = null;
          resolve(nodes);
        },
      );
    });
  };

  componentDidUpdate(prevProps) {
    const editor = this.editorRef.current;
    if (
      this.props.value.data !== prevProps.value.data &&
      !this.firstRenderWithEditorRef
    ) {
      this.convertHTMLToNodes(editor, this.props.value.data).then(() => {});
    }
  }

  componentDidMount() {
    this.convertHTMLToNodes(this.editorRef.current, this.props.value.data).then(
      (nodes) => {
        if (this.unmounting) return;

        this.setState((state, props) => {
          return { ...state, value: nodes };
        });
      },
    );
  }

  componentWillUnmount() {
    this.unmounting = true;
    this.setState(null);
  }

  firstRenderWithEditorRef = true;

  /**
   * Render method.
   * @method render
   * @returns {React.ReactElement} Markup for the component.
   */
  render() {
    const { id, className, properties, placeholder } = this.props;

    const withBlockProperties = (editor) => {
      editor.getBlockProps = () => this.props;
      return editor;
    };

    return (
      <FormFieldWrapper
        {...this.props}
        draggable={false}
        className="slate_wysiwyg"
      >
        <div
          style={{ boxSizing: 'initial' }}
          role="textbox"
          tabIndex="-1"
          onClick={() => {
            this.setState({ selected: true });
          }}
          className="slate_wysiwyg_box"
          onKeyDown={() => {}}
        >
          <SlateEditor
            id={id}
            className={className}
            name={id}
            value={this.state.value}
            block={`block-${id}`}
            renderExtensions={[withBlockProperties]}
            selected={this.state.selected}
            properties={properties}
            placeholder={placeholder}
            testingEditorRef={(val) => {
              this.editorRef.current = val;

              if (
                this.firstRenderWithEditorRef &&
                (__SERVER__ ? false : this.editorRef.current)
              ) {
                this.firstRenderWithEditorRef = false;
                this.convertHTMLToNodes(
                  this.editorRef.current,
                  this.props.value.data,
                ).then((nodes) => {
                  if (this.unmounting) return { ...state };

                  this.setState((state, props) => {
                    return { ...state, value: nodes };
                  });
                });
              }
            }}
            onChange={this.onChange}
          />
        </div>
      </FormFieldWrapper>
    );
  }
}

export default injectIntl(WysiwygWidget);
