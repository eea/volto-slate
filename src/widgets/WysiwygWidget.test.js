import React from 'react';
import WysiwygWidget from './WysiwygWidget';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import { fireEvent, render, screen } from '@testing-library/react';

const mockStore = configureStore();

global.__SERVER__ = true; // eslint-disable-line no-underscore-dangle
global.__CLIENT__ = false; // eslint-disable-line no-underscore-dangle

jest.mock('~/config', () => {
  const createEmptyParagraph = () => {
    return {
      type: 'p',
      children: [{ text: '' }],
    };
  };
  return {
    settings: {
      supportedLanguages: [],
      slate: {
        elements: {
          default: ({ attributes, children }) => (
            <span {...attributes}>{children}</span>
          ),
          strong: ({ attributes, children }) => (
            <strong {...attributes}>{children}</strong>
          ),
        },
        leafs: {},
        defaultBlockType: 'p',
        textblockExtensions: [],
        extensions: [],
        defaultValue: () => {
          return [createEmptyParagraph()];
        },
        persistentHelpers: [],
        contextToolbarButtons: [],
      },
    },
  };
});

window.getSelection = jest.fn();

test('renders a WysiwygWidget component', () => {
  const store = mockStore({
    intl: {
      locale: 'en',
      messages: {},
    },
  });
  const { asFragment } = render(
    <Provider store={store}>
      <WysiwygWidget
        id="qwertyu"
        title="My Widget"
        description="My little description."
        required={true}
        value={{ data: 'first <strong>render</strong>' }}
        onChange={() => {}}
      />
    </Provider>,
  );
  const f = asFragment();
  expect(f).toMatchSnapshot();
});

test('renders a WysiwygWidget component with working onChange prop', async () => {
  const onChangeMock = jest.fn();

  const store = mockStore({
    intl: {
      locale: 'en',
      messages: {},
    },
  });

  let v = { data: 'second <strong>render</strong>' };
  const { asFragment } = render(
    <Provider store={store}>
      <WysiwygWidget
        id="qwertyu"
        title="My Widget"
        description="My little description."
        required={true}
        value={v}
        onChange={onChangeMock}
      />
    </Provider>,
  );

  const f = asFragment();
  expect(f).toMatchSnapshot();

  const thing = await screen.findByText(/render/);

  fireEvent.blur(thing, {
    target: { innerHTML: 'renderA' },
  });

  await screen.findByText(/A/);

  const f2 = asFragment();
  expect(f2).toMatchSnapshot();

  // TODO: repair: this always fails:
  expect(onChangeMock).toHaveBeenCalledTimes(1);
});
