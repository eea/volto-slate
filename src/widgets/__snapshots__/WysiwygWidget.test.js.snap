// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a WysiwygWidget component 1`] = `
<DocumentFragment>
  <div
    class="inline required field help slate_wysiwyg"
    id="field-qwertyu"
  >
    <div
      class="ui grid"
    >
      <div
        class="stretched row"
      >
        <div
          class="four wide column"
        >
          <div
            class="wrapper"
          >
            <label
              for="field-qwertyu"
            >
              My Widget
            </label>
          </div>
        </div>
        <div
          class="eight wide column"
        >
          <div
            class="slate_wysiwyg_box"
            role="textbox"
            style="box-sizing: initial;"
            tabindex="-1"
          >
            <div
              class="slate-editor"
            >
              <div
                contenteditable="true"
                data-gramm="false"
                data-slate-editor="true"
                data-slate-node="value"
                role="textbox"
                spellcheck="false"
                style="outline: none; white-space: pre-wrap; word-wrap: break-word;"
              >
                <span
                  data-slate-node="element"
                >
                  <span
                    data-slate-node="text"
                  >
                    <span
                      class=""
                      data-slate-leaf="true"
                    >
                      <span
                        data-slate-string="true"
                      >
                        first 
                      </span>
                    </span>
                  </span>
                </span>
                <strong
                  data-slate-node="element"
                >
                  <span
                    data-slate-node="text"
                  >
                    <span
                      class=""
                      data-slate-leaf="true"
                    >
                      <span
                        data-slate-string="true"
                      >
                        render
                      </span>
                    </span>
                  </span>
                </strong>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="stretched row"
      >
        <div
          class="stretched twelve wide column"
        >
          <p
            class="help"
          >
            My little description.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders a WysiwygWidget component with working onChange prop 1`] = `
<DocumentFragment>
  <div
    class="inline required field help slate_wysiwyg"
    id="field-qwertyu"
  >
    <div
      class="ui grid"
    >
      <div
        class="stretched row"
      >
        <div
          class="four wide column"
        >
          <div
            class="wrapper"
          >
            <label
              for="field-qwertyu"
            >
              My Widget
            </label>
          </div>
        </div>
        <div
          class="eight wide column"
        >
          <div
            class="slate_wysiwyg_box"
            role="textbox"
            style="box-sizing: initial;"
            tabindex="-1"
          >
            <div
              class="slate-editor"
            >
              <div
                contenteditable="true"
                data-gramm="false"
                data-slate-editor="true"
                data-slate-node="value"
                role="textbox"
                spellcheck="false"
                style="outline: none; white-space: pre-wrap; word-wrap: break-word;"
              >
                <span
                  data-slate-node="element"
                >
                  <span
                    data-slate-node="text"
                  >
                    <span
                      class=""
                      data-slate-leaf="true"
                    >
                      <span
                        data-slate-string="true"
                      >
                        second 
                      </span>
                    </span>
                  </span>
                </span>
                <strong
                  data-slate-node="element"
                >
                  <span
                    data-slate-node="text"
                  >
                    <span
                      class=""
                      data-slate-leaf="true"
                    >
                      <span
                        data-slate-string="true"
                      >
                        render
                      </span>
                    </span>
                  </span>
                </strong>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="stretched row"
      >
        <div
          class="stretched twelve wide column"
        >
          <p
            class="help"
          >
            My little description.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders a WysiwygWidget component with working onChange prop 2`] = `
<DocumentFragment>
  <div
    class="inline required field help slate_wysiwyg"
    id="field-qwertyu"
  >
    <div
      class="ui grid"
    >
      <div
        class="stretched row"
      >
        <div
          class="four wide column"
        >
          <div
            class="wrapper"
          >
            <label
              for="field-qwertyu"
            >
              My Widget
            </label>
          </div>
        </div>
        <div
          class="eight wide column"
        >
          <div
            class="slate_wysiwyg_box"
            role="textbox"
            style="box-sizing: initial;"
            tabindex="-1"
          >
            <div
              class="slate-editor"
            >
              <div
                contenteditable="true"
                data-gramm="false"
                data-slate-editor="true"
                data-slate-node="value"
                role="textbox"
                spellcheck="false"
                style="outline: none; white-space: pre-wrap; word-wrap: break-word;"
              >
                <span
                  data-slate-node="element"
                >
                  <span
                    data-slate-node="text"
                  >
                    <span
                      class=""
                      data-slate-leaf="true"
                    >
                      <span
                        data-slate-string="true"
                      >
                        second 
                      </span>
                    </span>
                  </span>
                </span>
                <strong
                  data-slate-node="element"
                >
                  <span
                    data-slate-node="text"
                  >
                    <span
                      class=""
                      data-slate-leaf="true"
                    >
                      <span
                        data-slate-string="true"
                      >
                        renderA
                      </span>
                    </span>
                  </span>
                </strong>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="stretched row"
      >
        <div
          class="stretched twelve wide column"
        >
          <p
            class="help"
          >
            My little description.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
